bitswap/server/internal/decision/engine_test.go

```diff
@@ -1733,3 +1734,132 @@ func TestKillConnectionForInlineCid(t *testing.T) {
 		t.Fatal("connection was not killed when receiving inline in cancel")
 	}
 }
+
+func TestWantlistOverflow(t *testing.T) {
```

Comment:
This is good. Although I wonder if it'd also be good to have an integration test showing how things work (and when they slow down) as a result of basically not having backpressure.

It might make it easier for people to understand what happens if they run into issues here and for us to explain some motivations behind protocol change proposals.

Doesn't seem strictly necessary but would be good to have.

Reply:
Agree.

The whole problem stems from only removing wants when [blocks](https://github.com/ipfs/boxo/blob/main/bitswap/server/internal/decision/engine.go#L960) or [block presence](https://github.com/ipfs/boxo/blob/main/bitswap/server/internal/decision/engine.go#L966) has been sent. That, and the somewhat incorrect (IMO) [truncation of the incoming wants](https://github.com/ipfs/boxo/blob/main/bitswap/server/internal/decision/engine.go#L705-L714) based on the remaining space left in the message queue / peer ledger. An integration test should only have to send a couple messages where the DONT_HAVE blocks add up to the queue size limit,  and before this PR bitswap should stop working, and after this PR it should continue working.